### PR TITLE
update to cache-core

### PR DESCRIPTION
**Pull request by jbaxleyiii**
@nevir @yuit this updates to the latest stable API for the Apollo Cache as well as factors out some copied utilities in favor of the new apollo-utilities package on npm.

One thing to note is the name change for the main ApolloCache interface. Since the cache-core is ApolloCache, I thought it was best to name this implementation Hermes instead of overwriting the class namespace. Thoughts?

I'm still figuring my way around so please let me know thoughts!

### CODE REVIEW

**Comment by jbaxleyiii** [comment]
@nevir I'm going to update the package name here (apollo-cache-core => apollo-cache) 👍

File: src/apollo/Queryable.ts [hunk 1]
```diff
@@ -1,28 +1,34 @@
+/* eslint-disable class-methods-use-this */
```
Comment (nevir) [inline src/apollo/Queryable.ts]: You heathen!  break 'em out into helper funcs, or just disable it for the specific method

File: src/apollo/Queryable.ts [hunk 2]
```diff
@@ -1,28 +1,34 @@
...
 /**
  * Apollo-specific interface to the cache.
  */
-export abstract class ApolloQueryable {
+export abstract class ApolloQueryable implements DataProxy {
   /** The underlying Hermes cache. */
   protected abstract _queryable: Queryable;
 
-  diffQuery(options: interfaces.Cache.DiffQueryOptions): interfaces.Cache.DiffResult {
+  diff<T = any>(options: Cache.DiffOptions): Cache.DiffResult<T | any> {
```
Comment (nevir) [inline src/apollo/Queryable.ts]: Hmm, why `T | any` for the result? (particularly if the template parameter already defaults to `any`)

File: src/apollo/Queryable.ts [hunk 3]
```diff
@@ -1,28 +1,34 @@
...
     const query = toQuery(options.query, options.variables);
     const { result, complete } = this._queryable.read(query, options.optimistic);
     if (options.returnPartialData === false && !complete) {
       // TODO: Include more detail with this error.
       throw new Error(`diffQuery not satisfied by the cache.`);
     }
 
-    return { result, isMissing: !complete };
+    return { result, complete };
```
Comment (nevir) [inline src/apollo/Queryable.ts]: woot!

File: src/apollo/index.ts [hunk 4]
```diff
@@ -1 +1,4 @@
-export { ApolloCache } from './ApolloCache';
+import { Hermes } from './Hermes';
+export { Hermes };
+export default Hermes;
```
Comment (nevir) [inline src/apollo/index.ts]: nit: let's drop the export `default` (particularly for private modules like this) - rationale: `default` exports construct a new symbol to reference that value, making it difficult to refactor the name of something

File: src/util/ast.ts [hunk 5]
```diff
@@ -1,47 +1,21 @@
...
-
-  return names;
-}
+export {
+  getOperationDefinitionOrDie as getOperationOrDie,
+  variablesInOperation,
+  valueFromNode,
+  FragmentMap,
+  addTypenameToDocument,
+} from 'apollo-utilities';
```
Comment (nevir) [inline src/util/ast.ts]: glorious!

File: src/Cache.ts [hunk 6]
```diff
@@ -57,7 +57,7 @@ export class Cache implements Queryable {
    * Registers a callback that should be triggered any time the nodes selected
    * by a particular query have changed.
    */
-  watch(query: Query, callback: () => void): () => void {
+  watch(query: Query, callback: (newData: any) => void): () => void {
```
[the remark this answers is not in the document]
Comment (jbaxleyiii) [reply src/Cache.ts]: agreed! I *think* I added that to the latest release. If not I'll do it!

File: scripts/test-unit-coverage.sh [hunk 7]
```diff
@@ -13,7 +13,7 @@ fi
 run test:unit -- --coverage "${OPTIONS[@]}"
 
 if [[ "${CI}" == "" ]]; then
-  open ./output/test:unit/index.html
+  open ./output/test-unit/index.html
```
[the remark this answers is not in the document]
Comment (jbaxleyiii) [reply scripts/test-unit-coverage.sh]: sure thing!

File: src/apollo/Hermes.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  /** The underlying Hermes cache. */
+  protected _queryable: Cache;
+
+  constructor(configuration?: CacheContext, intitialState?: GraphSnapshot) {
+    super();
+    this._queryable = new Cache(configuration, intitialState);
+  }
+
+  getData(): any {
+    return this._queryable.getSnapshot().baseline;
```
[the remark this answers is not in the document]
Comment (jbaxleyiii) [reply src/apollo/Hermes.ts]: `getOptimisticData ` is currenty used by the dev tools. `getData` was used for getting the store for SSR. Honestly I'll going to try to kill both of them and maybe make an optional API for SSR? `.extract()`?

However, for SSR, having a way to serialize and boot up from will be critical for a number of users. Thoughts on how I should go about implementing it?

[the remark this answers is not in the document]
Comment (nevir) [reply src/apollo/Hermes.ts]: Ah!  I think for that, we can have a unified API for cache implementers - I think something like:

```
interface ApolloCache<TSerialized> {
  /**
   * Replaces existing state in the cache (if any) with the values expressed by
   * `serializedState`.
   *
   * Called when hydrating a cache (server side rendering, or offline storage),
   * and also (potentially) during hot reloads.
   */
  restore(serializedState: TSerialized): void;

  /**
   * Exposes the cache's complete state, in a serializable format for later restoration.
   */
  save(optimistic?: boolean): TSerialized;
}
```

**Pull request merged**